Task 7

ruslanrix opened the pull request.

ruslanrix committed:
Add files via upload

ruslanrix committed:
Add files via upload

zverok commented on movielist_class.rb:
```diff
@@ -81,4 +81,21 @@ def m_sort_by_reverse (input)
+  end
+
+  def add_sort_algo (arg, &block)
+    @sort_algo_collect[arg] = proc { |m| yield m if block_given? }
```
```ruby
@sort_algo_collect[arg] = block
```
;)

ruslanrix replied:
О! А я так делал еще до возникшей мысли с хэшом блоков, но потом увлекся proc и забыл)

zverok commented on movielist_class.rb:
```diff
@@ -81,4 +81,21 @@ def m_sort_by_reverse (input)
+  end
+
+  def sort_by(arg)
+    @movielist.sort_by(&@sort_algo_collect[arg]).map { |m| puts "#{m.movie} : #{m.genre} : #{m.year} : #{m.director}" }
```
Ну да, только puts отсюда убрать (если ты его не для отладки добавлял).
Ну и обработать вариант «мы не знаем такого алгоритма»

ruslanrix committed:
Add files via upload

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
   end
 
+  def print
+     @movielist.each{|m| puts yield m if block_given?  }
```
Хорошо бы чтобы если блок не передан, выводилось какое-то представление фильма по умолчанию, а не происходило ничто.

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+  end
+
+  def sorted_by
+    @movielist.sort_by{ |m| yield m if block_given? }.map { |m| "#{m.movie} : #{m.genre.join(', ')} : #{m.year}" }
```
Не, лучше пусть возвращает фильмы, а не какие-то форматированные строки.

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+  end
+
+  def sort_by(arg)
+    if @sort_algo_collect[arg] != nil
```
не обязательно сравнивать с `nil` явно, можно просто `if @sort_algo_collect[arg]`

Но лучше всего в таком порядке:
* вытащить в переменную алгоритм
* бросить ошибку (raise) если он не найден
* отсортировать

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+
+  def sort_by(arg)
+    if @sort_algo_collect[arg] != nil
+    @movielist.sort_by(&@sort_algo_collect[arg]).map { |m| "#{m.movie} : #{m.genre.join(', ')} : #{m.year} : #{m.director}" }
```
опять же, лучше фильмы вернуть, а не какие-то строки

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+  end
+
+  def filter(arg)
+    if arg.any? { |k, v| @filters[k] == nil}
```
лучше проверять через `filters.key?(k)`

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+
+  def filter(arg)
+    if arg.any? { |k, v| @filters[k] == nil}
+      "Some of the filters: #{arg.keys.join(', ')} - is unknown"
```
1. лучше кидать ошибку (raise)
2. лучше точно указать, какие фильтры неизвестны

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+      else
+        super
+      end
+  end
```
Какой-то кривоватый ход. А зачем?

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,61 @@ def actor_stats
+    if arg.any? { |k, v| @filters[k] == nil}
+      "Some of the filters: #{arg.keys.join(', ')} - is unknown"
+    else
+    @movielist.select{ |m| arg.all? { |k, v| @filters[k].call(m, *v) } }
```
Лучше возвращать список фильмов.

zverok reviewed (changes requested):
Неплохо. Но надо бы почистить!

ruslanrix committed:
Add files via upload

  def include?(parameter)
      (parameter.is_a? Array) ? (return parameter.any?{|x| self.include? x }) : super
  end
Если правильно понял, в одной строчке записал. А вот если без "super", тогда валится фильтрация.

Ну и еще один момент. Мне показалось лишним добавлять присваивание алгоритма к переменной в def sort_by(arg).Так, вроде же нормально?

ruslanrix commented:
```
def include?(parameter)
      (parameter.is_a? Array) ? (return parameter.any?{|x| self.include? x }) : super
  end
```
Если правильно понял, в одной строчке записал. А вот если без "super", тогда валится фильтрация.

Ну и еще один момент. Мне показалось лишним добавлять присваивание алгоритма к переменной в def sort_by(arg).Так, вроде же нормально?


1. Блок кода - это произвольный код, передаваемый любому методу в качестве неявного последнего аргумента.
2. Замыкание  - это блок или выражение, заключенное внутри {}, если оно однострочное, или выражение между do ... end.

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,57 @@ def actor_stats
+  end
+
+  def sorted_by
+    @movielist.sort_by{ |m| yield m if block_given? }.map { |m| m.movie }
```
Во-первых, можно же `map(&:movie)`, а во вторых — зачем вообще? Почему не возвращать весь объект-фильм?

ruslanrix replied:
Насчет map(&:movie) да, точно. 
Ааа, то есть под фильмом подразумевается объект. А если `map(&:movie)`, то будет выводиться список наименований фильмов. Теперь понял, а то нижеследующий образец вывода списка фильмов мне казался странным.
```
#<NewMovie:0x000000014fd638>
#<ModernMovie:0x000000011ae9f0>
#<ModernMovie:0x00000001389c20>
#<NewMovie:0x0000000158a858>
```

zverok replied:
Переопредели метод `to_s` у фильма, будет выводиться любая красота на которую фантазии хватит.

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,57 @@ def actor_stats
+  end
+
+  def filter(arg)
+    @unknown = []
```
Почему `@`-переменная, она где-то ещё будет использоваться?

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,57 @@ def actor_stats
+
+  def filter(arg)
+    @unknown = []
+    if arg.keys.each { |k| @unknown.push(k) if !@filters.key?(k) }.all? {|k| @unknown.empty? }
```
Что-то какое-то дико запутанное условие, как его читать?..

ruslanrix replied:
`if arg.keys.all? { |k| @filters.key?(k) ? true : (unknown.push(k); false) }`

А если так, то получается следующее: в случае двух ошибочных фильтров, в ошибке будет указываться только самый первый ошибочный фильтр, но не все сразу. Или это верно - последовательное выкидывание ошибок?

zverok replied:
Не увлекайся так однострочностью, это не всегда позитив.

Я бы написал так:
```ruby
unknown = args.keys.reject { |k| filters.key?(k) }
fail .... unless unknown.empty?
# ... а здесь уже нормальная обработка фильтров без лишних if-ов
```

Задач-то тут две — значит так и надо писать:
* заругаться, если фильтры неизвестны
* обработать известные фильтры

zverok reviewed (changes requested):
Работать работает, а качество кода надо бы улучшить!

ruslanrix committed:
Add files via upload

ruslanrix committed:
Add files via upload

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,55 @@ def actor_stats
+  end
+
+  def filter(arg)
+    unknown = []
```
А эта строчка зачем?

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,55 @@ def actor_stats
+    unknown = []
+    unknown = arg.keys.reject { |k| @filters.key?(k) }
+    fail "The following filters: #{unknown.join(', ')} - are unknown" unless unknown.empty?
+    @movielist.select{ |m| arg.reject {|a| unknown.include?(a) }
```
А зачем этот reject? Если у нас какие-то фильтры были неизвестны, уже всё равно ошибка улетела.

ruslanrix replied:
Аа, у меня была мысль - выводить и ошибку, и результат обработки уже толькоизвестных фильтров выводить. Но единственный способ, который я нашел, это просто вместо _fail/raise_ поставить _p_.
Потом понял, что есть вроде изощренные способы, но в данном случае правильнее - выводить либо ошибку, либо результат.

zverok commented on movielist_class.rb:
```diff
@@ -68,17 +68,55 @@ def actor_stats
+class Array
+  def include?(parameter)
+      (parameter.is_a? Array) ? (return parameter.any?{|x| self.include? x }) : super
+  end
```
Ну, не хорошо это, для своих мелких целей изменять поведение базового класса. Придумай другое решение.

ruslanrix replied:
А да, упустил это. Причем когда to_s переопределял, то была мысль и про .include

zverok reviewed (changes requested):
В принципе уже почти!
